language/move-lang/src/parser/byte_string.rs

```diff
@@ -0,0 +1,304 @@
+    }
+
+    #[test]
+    fn test_non_ascii_characters() {
+        check_dec_error(
+            r"Only ascii characters are supported. ♎.",
+            37,
+            37,
+            "Invalid character ‘♎’. Only ascii characters are supported.\n\
+                     For Unicode characters use \\u{…} notation.",
+        );
+
+        check_dec_error(
+            r"Оnly ascii characters are supported.",
+            0,
+            0,
+            "Invalid character ‘О’. Only ascii characters are supported.\n\
+                     For Unicode characters use \\u{…} notation.",
+        );
+    }
+
+    #[test]
+    fn test_success_ascii_escapes() {
+        assert_eq!(
+            dec(r"Hello\n libra. \n	Newline; \r	Carriage return; \t	Tab; \\	Backslash; \0 Null"),
+            b"Hello\n libra. \n	Newline; \r	Carriage return; \t	Tab; \\	Backslash; \0 Null".to_vec(),
+        );
+    }
+
+    #[test]
+    fn test_success_quotes() {
+        assert_eq!(dec("Hello \\\" "), b"Hello \" ".to_vec(),);
+    }
+
+    #[test]
+    fn test_success_hex() {
+        assert_eq!(
+            dec(r"Test string with hex [\x01\x02-\x03\x04\x05]."),
+            b"Test string with hex [\x01\x02-\x03\x04\x05].".to_vec(),
+        );
+    }
+
+    #[test]
+    fn test_unicode_sequence() {
+        assert_eq!(dec(r"\u{264E}."), "♎.".as_bytes().to_vec(),);
+
+        assert_eq!(dec(r"T\u{EB}."), "T\u{EB}.".as_bytes().to_vec(),);
+
+        assert_eq!(dec(r"\u{10FFFF}."), "\u{10FFFF}.".as_bytes().to_vec(),);
+    }
+
+    #[test]
+    fn test_invalid_hex_sequence() {
+        check_dec_error(r"\x0y", 3, 3, "Invalid hexadecimal character: 'y'");
+
+        check_dec_error(
+            r"\x0",
+            0,
+            3,
+            "Invalid escape: '\\x0'. Single HEX literal must be represented by two symbols. [\\x00-\\xFF].",
+        );
+
+        check_dec_error(r"\x{0", 2, 2, "Invalid hexadecimal character: '{'");
+
+        check_dec_error(r"test \x{0", 7, 7, "Invalid hexadecimal character: '{'");
+    }
+
+    #[test]
+    fn test_invalid_unicode_sequence() {
+        check_dec_error(r"\u00ffffff}", 0, 3, "Invalid Unicode escape: '\\u0'");
+
+        check_dec_error(
+            r"\u{00ffffff}",
+            0,
+            10,
+            "Unicode escape must be at most 10FFFF",
+        );
+
+        check_dec_error(
+            r"\u",
+            0,
+            2,
+            "Invalid escape: '\\u'. Unicode escape should be in \\u{...} format.",
+        );
+
+        check_dec_error(
+            r"\u{",
+            0,
+            3,
+            "Invalid escape: '\\u{'. Unicode escape should be in \\u{...} format.",
+        );
+
+        check_dec_error(
+            r"\u{00",
+            0,
+            5,
+            "Invalid escape: '\\u{00'. Unicode escape should be in \\u{...} format.",
+        );
+
+        check_dec_error(r"\u{io}", 0, 5, "Invalid Unicode escape: '\\u{io}'");
+    }
```

Comment:
Could you move all of these to functional tests?

Reply:
Do you mean functional tests on move lang? https://github.com/libra/libra/tree/master/language/move-lang/tests/functional

Reply:
Done.